Fix for OSMWidget in formsets

blag opened the pull request:
Fixes #7.

Got the idea from [this StackOverflow post](http://stackoverflow.com/a/8272261).

MarkusH commented:
I feel like there has to be another way to get the prefix out of a form in a formset. You might want to look at `django.forms` for how it's implemented there.

blag commented:
I reworked this PR so it no longer requires any customization of the form itself and rebased it on #11 and #8.

I created a `OSMFormField` in `forms.py` and moved `OSMWidget` into `widgets.py`, so imports will need to be changed for users upgrading to this. I can add reverse-compatibility shims into it if necessary.

MarkusH commented on osm_field/fields.py:
```diff
@@ -200,12 +201,14 @@ def deconstruct(self):
-        widget = OSMWidget(lat_field=self.latitude_field_name,
+        widget = OSMWidget(
+            lat_field=self.latitude_field_name,
             lon_field=self.longitude_field_name)
```
Can you add a tailing comma before the `)` and put the `)` on a new line, please.

```python
widget = OSMWidget(
    lat_field=...,
    lon_field=...,
)
```

blag commented:
Done.

blag commented:
As far as I've been able to find, for django < 1.9 there's no way to pass a form's prefix (eg: in a formset) to each widget individually. To work around this, I created a form mixin that adds the form's prefix to each `OSMFormField` field:

```python
class OSMFormMixin(object):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if django.VERSION < (1, 9):
            for f in self.fields:
                if issubclass(f, OSMFormField):
                    f.widget.attrs['prefix'] = self.prefix
```

This will also work for Django >= 1.9, but it's unnecessary. I have updated the usage documentation to reflect this.

blag commented:
Okay, reworked everything:

* Took care of merge conflicts
* Added tests for new code
* Noticed that Django < 1.9 was failing, so I fixed that
* Fixed new code flake8 and isort tests
* Updated documentation

So this changeset looks ready to go as far as I can tell. Let me know if you would like me to do anything else for it. Thanks!

MarkusH commented on tests/migrations/0001_initial.py:
```diff
@@ -64,4 +64,27 @@ class Migration(migrations.Migration):
             },
             bases=(models.Model,),
         ),
+        migrations.CreateModel(
```
Please don't alter existing migrations and add a new one instead.

blag replied:
Usually I would agree with you, and I would never have done that for the project's main migrations. But those are the migrations for the test fixtures - that database is setup *from scratch* for every test and completely torn down afterward. Because of that, nobody should have any of those objects in their persistent database, so modifying that migration won't/shouldn't impact any users. Having multiple migrations for the test database is, as far as I know, pointless.

Let me know if you still want me to create a new migration instead - I'm more interested in getting this merged.

MarkusH replied:
Oops, you're right. Makes sense for the tests. Thanks. I'll have a look
at the PR tomorrow. Thank you for your contribution!

On Mon, Jun 06, 2016 at 11:23:49AM -0700, blag wrote:
>> @@ -64,4 +64,27 @@ class Migration(migrations.Migration):
>>              },
>>              bases=(models.Model,),
>>          ),
>> +        migrations.CreateModel(
>
>Usually I would agree with you, and I would never have done that for the project's main migrations. But those are the migrations for the test fixtures - that database is setup *from scratch* for every test and completely torn down afterward. Because of that, nobody should have any of those objects in their persistent database, so modifying that migration won't/shouldn't impact any users. Having multiple migrations for the test database is, as far as I know, pointless.
>
>Let me know if you still want me to create a new migration instead - I'm more interested in getting this merged.
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/MarkusH/django-osm-field/pull/9/files/4821f8b19435808048b1735568f11f713a1f1bbd#r65941875

blag replied:
No worries! And I don't know if you caught it or not, but I fixed the code coverage loss (this PR now increases coverage by 1.96%) - @codecov-io's [comment](https://github.com/MarkusH/django-osm-field/pull/9#issuecomment-223791232) is out of date at this point in time.

blag commented:
Bump. Let me know if there's anything else you'd like me to do for this PR.

And just FYI: I plan on rebasing my other PRs off of master as you merge their preceding PRs in to minimize their diffs and ease review. :smile:
